sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/sharedconfig/build.sbt

```diff
@@ -51,7 +49,9 @@ InputKey[Unit]("html") := {
 }
 
 TaskKey[Unit]("checkSize") := {
-  val size = IO.readBytes((webpack in (Compile, fullOptJS)).value.head).length
+  val files = (webpack in (Compile, fullOptJS)).value
+  val bundleFile = files.find(_.getName().contains("bundle")).get
```

Comment:
In the `LibraryAndApplication` mode the `webpack` task does not return only the bundle?

Reply:
Correct, it returns the library, loader, scalajs AND the bundle. This allows your to turn it on and perhaps load the -bundle on some pages but use the individual pieces elsewhere.

Reply:
I think that’s a very fragile way to return multiple data to users. In case of the `LibraryAndApplication` mode, I would expect to get only one result from webpack: the bundle. If I want to use the individuaul pieces then I would use the `LibraryOnly` mode where I would expect the `webpack` task to return the libraries bundle and the `fastOptJS` or `fullOptJS` task to return the application.

Reply:
If users think it’s useful to return all these files in the `LibraryAndApplication` mode, then we should probably wrap them into data types that describe their meaning:

~~~ scala
sealed trait BundlingArtifact { def file: File }
case class Bundle(file: File) extends BundlingArtifact
case class ScalaJSOutput(file: File) extends BundlingArtifact
case class Loader(file: File) extends BundlingArtifact
~~~

But IMHO it makes more sense to only return the final bundle.

Reply:
Well, tbh, that's how I had it, but folks didn't want to add so many additional keys, which is necessary in order to wire everything into sbt-web. As things stand, the sbt-web plugin just connects up the outputs from the webpack task and doesn't need to know anything else.

In LibraryOnly mode we have to return at least two files (the loader and the library), and thats assuming the -fastopt file gets carried along elsewhere.

Reply:
One possible solution to this dilemma is to use *attributes*. For example we use that to attach the location the source map file associated to a .js file: https://github.com/scala-js/scala-js/blob/19742d5e2b720934c75ec5089bb3c5260d8ed9e9/sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala#L187-L188

Reply:
I don't personally care much if `LibraryAndApplication()` returns only the bundle file, I don't intend to use it anyway. But, I can say, having been using this for several weeks, having to change both my frontend references and the bundling mode when switching back and forth between modes is inconvenient. The -loader, -library, and -fastopt are all sitting there to be copied over, so why not make them available?

In general, I think `LibraryAndApplication()` is mostly a transitional mode. It's a drop-in replacement for `Application`, and starts making the assets available for `LibraryOnly` with most (but not all) of the performance benefits. 

If the files are there, folks can transition over to `LibraryOnly` at their own pace, only changing their frontend references. If not, it's all or nothing and an SBT restart between.

In general, if we're seriously going to attack this, I think we need to seriously consider why we are carrying around `Seq[(String, File)]` in the first place. I don't think we have a test-case that exercises it and it introduces a ton of extra complexity.

Reply:
I *was* wondering what all those `Attributed[File]` references were for! That does seem like a good, less-brittle solution. At it's simplest, we could leave everything as is and introduce the attributes, one for each type of file, and move on.

I'd still like to understand if theres actually a valid use-case for the multiple entry points, because the entire stack otherwise be `[Task[Seq[Attributed[File]]]`.

In other words, instead of 
```
  private[sbtplugin] def entries(stage: TaskKey[Attributed[File]])
    : Def.Initialize[Task[Seq[(String, File)]]] =
    Def.task {
      val stageFile = stage.value.data
      val name = stageFile.name.stripSuffix(".js")
      Seq(name -> stageFile)
    }
```

Why couldn't we just do this:
```
  private[sbtplugin] def entries(stage: TaskKey[Attributed[File]])
    : Def.Initialize[Task[Attributed[File]] =
    Def.task {
      stage.value.data
    }

private[sbtplugin] def webpack(
      stage: TaskKey[Attributed[File]]): Def.Initialize[Task[Attributed[File]]] = ???

```

In general, it seems like we've been designing back from the idea we might bundle multiple scala.js entry points into the same webpack bundle, but in order to do that we have `lift` most of that logic up a level anyway, to the sbt-web level where we have several scala.js output projects.

Reply:
> To that end, @sjrd, does the scala.js plugin already provide an attribute with the "project name"?

No. The file I pointed to earlier shows everything that Scala.js provides.

Reply:
> If the files are there, folks can transition over to LibraryOnly at their own pace, only changing their frontend references. If not, it's all or nothing and an SBT restart between.

Fair enough.

Also, I’m ok with using `Attributed[File]` as a result, but then we need to clearly document what “attributes” we are going to attach to our results so that users easily know how to filter out the files that they are not interested in.

> In general, it seems like we've been designing back from the idea we might bundle multiple scala.js entry points into the same webpack bundle, but in order to do that we have to "lift" most of that logic up a level anyway, to the sbt-web level where we have several scala.js output projects.

That’s right. Initially I thought it might be easy to use several entry points but it turned out it’s not. Since anyway you removed the `webpackEntries` key, you can also do the simplification you mentioned in your comment.